pkg/obfuscate/appsec.go

```diff
@@ -0,0 +1,538 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package obfuscate
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+// ObfuscateAppSec obfuscates the given appsec tag value in order to remove
+// sensitive values from the appsec security events.
+func (o *Obfuscator) ObfuscateAppSec(val string) string {
+	keyRE := o.opts.AppSec.ParameterKeyRegexp
+	valueRE := o.opts.AppSec.ParameterValueRegexp
+	if keyRE == nil && valueRE == nil {
+		return val
+	}
+	ob := appsecEventsObfuscator{
+		keyRE:   keyRE,
+		valueRE: valueRE,
+	}
+	output, err := ob.obfuscate(val)
+	if err != nil {
+		o.log.Errorf("unexpected error while obfuscating the appsec events: %v", err)
+		return val
+	}
+	return output
+}
+
+type appsecEventsObfuscator struct {
+	keyRE, valueRE *regexp.Regexp
+}
+
+type unexpectedScannerOpError int
+
+func (err unexpectedScannerOpError) Error() string {
+	return fmt.Sprintf("unexpected json scanner operation %d", err)
+}
+
+// inputDiff holds the list of diffs, built by the obfuscator, and to ultimately
+// apply to the input.
+type inputDiff []struct {
+	from, to int
+	value    string
+}
+
+// Add a new entry to the list. Note that due to how the apply() method works
+// by creating the output string by walking this list, the list entries must be
+// sorted in ascending "input offset" order.
+func (d *inputDiff) add(from, to int, value string) {
+	elt := struct {
+		from  int
+		to    int
+		value string
+	}{
+		from:  from,
+		to:    to,
+		value: value,
+	}
+	diff := *d
+	l := len(diff)
+	i := sort.Search(l, func(i int) bool {
+		return (*d)[i].to > from
+	})
+	if i == l {
+		*d = append(diff, elt)
+		return
+	}
+	*d = append(diff[:i+1], diff[i:]...)
+	(*d)[i] = elt
+}
+
+// Apply the diff to the input if any.
+func (d inputDiff) apply(input string) string {
+	from := 0
+	var output strings.Builder
+	for _, diff := range d {
+		output.WriteString(input[from:diff.from])
+		from = diff.to + 1
+		output.WriteString(diff.value)
+	}
+	output.WriteString(input[from:])
+	return output.String()
+}
+
+// Append the given diff to the current one, shifted by the given offset.
+// Cf. method obfuscateRuleMatchParameter()
+func (d *inputDiff) appendTo(diff *inputDiff, offset int) {
+	for _, d := range *d {
+		diff.add(d.from+offset, d.to+offset, d.value)
+	}
+}
+
+// Entrypoint of the appsec obfuscator based on a json scanner and which tries
+// to only obfuscate the appsec rule parameters of the given input by searching
+// and targeting objects of the form
+// `{ "parameters": <appsec rule match parameters>, <any other keys>: <any other values> }`
+// When such object is found, its parameters' value is obfuscated.
+func (o *appsecEventsObfuscator) obfuscate(input string) (output string, err error) {
+	var (
+		scanner scanner
+		diff    inputDiff
+	)
+	scanner.reset()
+	keyFrom := -1
+	keyTo := -1
```

Comment:
```suggestion
	keyFrom := -1 // marks the starting index of a JSON key in bytes
	keyTo := -1 // marks the end index of a JSON key in bytes
```
This is quite complex code, so let's make sure we're as helpful as possible.

Reply:
Done in https://github.com/DataDog/datadog-agent/pull/11324/commits/594fc9901e1c0b15fec3add5a4ff269c403e7abe

Reply:
Not done here 🙂